### DISCUSSION TITLE
on-demand-ephemeral-self-hosted-runners/

### DISCUSSION THREAD

**Discussion by giscus** [colindembovsky/colindembovsky.github.io#6] [Giscussions]
# on-demand-ephemeral-self-hosted-runners/

Do you need to deploy to private VNets using GitHub Actions, but don’t want to have to keep self-hosted runners running all the time? In this post I show you how you can use Ephemeral Runners to create on-demand self-hosted runners.

https://colinsalmcorner.com/on-demand-ephemeral-self-hosted-runners/

**Comment by DanyC97** [comment]
Very interesting approach , you mind sharing how long will it take for the `deploy_runner` job to complete and be ready for the next job to use it? 

I have s'thing similar where ASG (EC2) on which run n self-hosted gh containers and they always up hence avoiding the need to to wait to get it up and in online mode.

**Reply by colindembovsky** [reply]
Hey @DanyC97 it takes about 90 seconds or so to spin up. Depends on how busy Azure is at that time, and on how large the container image is that you're spinning up.

This approach isn't about minimizing wait time - it's about having a once-off (ephemeral) runner that can execute exactly one job. If you need blazing response times, then you probably want to use an auto-scaling option with a full k8s cluster that always has some agents ready and can scale up when those agents are busy. These are different scenarios.

**Comment by larryclaman** [comment]
Have you run into any issues when using this with container-based actions?  https://docs.github.com/en/actions/creating-actions/creating-a-docker-container-action

**Reply by colindembovsky** [reply]
@larryclaman I have not tried container-based actions, but that would require docker-in-docker. There may be some additional configuration you need to mount the docker socket, if you even can with ACI.

**Reply by joshjohanning** [reply]
@larryclaman @colindembovsky - while docker supports docker in docker, GitHub Actions doesn't support [container jobs](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idcontainer).

https://github.com/actions/runner/issues/406
https://github.com/actions/runner/issues/367#issuecomment-939454503

> Error: Container feature is not supported when runner is already running inside container.

![image](https://user-images.githubusercontent.com/19912012/153923443-dd86737a-585a-463f-af63-df453c5f460e.png)

**Reply by joshjohanning** [reply]
@larryclaman @colindembovsky oops what I have above is for _[container jobs](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idcontainer)_ - Docker actions will work, but you have to [volume map to the host Docker socket](https://stackoverflow.com/questions/44451859/how-to-install-docker-in-docker-container).

Something like this in the `docker run` command: `-v /var/run/docker.sock:/var/run/docker.sock`

Had to play with permissions to get that to work - `chmod 777 /var/run/docker.sock` worked albeit it's a little open. 

Note I did not try this in ACI though.
